### ISSUE TITLE
Enhance TaskContext and add task failure root cause

### ISSUE THREAD

**Issue by mingmwang** [apache/arrow-datafusion#3410]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

In current Datafusion/Ballista plan's execution path, the error result mapping is a chaos, the root cause of the task failure is not clear, especially when the plan tree contains pipeline breakers.  
To make the error handling clear and make it easy to reasoning about the real task failure reason,  the suggestion is to add a new method to TaskContext to set the task failure root cause into TaskContext.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by mingmwang** [comment]
@yahoNanJing

**Comment by alamb** [comment]
I think in general the wrapping of errors is definitely not valuable 

The datafusion code already tries to unwrap an ArrowError here (when converting from DataFusionError --> ArrowError)

https://github.com/apache/arrow-datafusion/blob/master/datafusion/common/src/error.rs#L209-L217

I wonder if we can do something similar in Ballista? Or maybe write some "unwrapper" that would strip out the useless wrapping errors before display.

**Comment by yahoNanJing** [comment]
Thanks Alamb, the context here is not just for displaying. Recently we are working on error handling and error recovering which is useful for distributed execution. And we hope to do error recovering for some special error case and we need to know the root cause of the error. If an error is wrapped too much, it will be very difficult to make decision of which error to recover and others not.

**Comment by alamb** [comment]
> And we hope to do error recovering for some special error case and we need to know the root cause of the error. If an error is wrapped too much, it will be very difficult to make decision of which error to recover and others not.

That is a neat usecase and it makes sense. 

Arrow and DataFusion don't have very specific errors (unlike some other rust crate where the specific error call site often has its own unique variant). This makes coding easier as there is less error machinery to maintain, but it makes distinguishing between errors potentially more challenging -- All you have is the handful of variants and a message.

As the rest of the rust ecosystem works on stabalizing "error context" (aka being able to walk up the chain of error messages) maybe finding the root cause will become simpler.

I mostly mention this as maybe it would be a good time to figure out if we can add a `source:` of each `ArrowError`, and `DataFusionError` as a more general purpose method than adding such info to the `TaskContext`

**Comment by yahoNanJing** [comment]
> I mostly mention this as maybe it would be a good time to figure out if we can add a source: of each ArrowError, and DataFusionError as a more general purpose method than adding such info to the TaskContext

It would be very helpful for error recovering. @mingmwang, what do you think?

**Comment by alamb** [comment]
> But it still depends on the programer to propagate the errors correctly in the execution path. And there is another case that we still could not catch the error easily.
> If we add such info to TaskContext, errors can be propagated easily and there is less burden to the programer.

Yes I agree care would be needed in various cases. I am not opposed to adding information to the TaskContext -- perhaps we can brainstorm additional improvements to errors / error handling as a follow on project. I know @tustvold has some ideas in https://github.com/apache/arrow-rs/pull/2711 etc so maybe it would be part of a broader ecosystem improvment

**Comment by tustvold** [comment]
> If we add such info to TaskContext, errors can be propagated easily and there is less burden to the programer.

Calling tokio::spawn without doing anything with the join handle is not a good idea, not only will you potentially silently drop errors, as you elude to, but more problematically you will miss panics. This has been a frequent pain point, see the work by @crepererum to fix this https://github.com/apache/arrow-datafusion/pulls?q=is%3Apr+is%3Aclosed+author%3Acrepererum. Longer term I'm hoping to move away from tokio see https://github.com/apache/arrow-datafusion/issues/2504, in part to do away with these types of problems, but I'm not sure when I'll get back to working on that. In the meantime, a tokio::spawn that doesn't stash the JoinHandle is an anti-pattern IMO. 

> additional improvements to errors / error handling

If we are to go down this route, I think the question to ask is "what users are there of the current structured error handling". In particular, could we just use something like [anyhow](https://docs.rs/anyhow/latest/anyhow/) which we already have a transitive dependency on as a result of prost...

**Comment by crepererum** [comment]
> Calling `tokio::spawn` without doing anything with the join handle is not a good idea, not only will you potentially silently drop errors, as you elude to, but more problematically you will miss panics.

It's not only about error handling though but also about task cancellation. Esp. in a server application, it's good to stop a query from running when the original request is cancelled. Uncaptured task handles will prevent this from happening, since tokio will still drive the execution.

**Comment by alamb** [comment]
> See https://github.com/rust-lang/rust/issues/58520 for work to stabilize a mechanism to "walk" the tree

Yes, this is what I really want out of errors in DataFusion and Arrow

**Comment by alamb** [comment]
Filed https://github.com/apache/arrow-rs/issues/2725 to discuss better error handling in arrow (and thus also datafusion and ballista)
